### PR TITLE
add [contenthash] support

### PR DESCRIPTION
**Pull request by sokra**
**What kind of change does this PR introduce?**
feature

**Did you add tests for your changes?**
a bit

**If relevant, link to documentation update:**
TODO

**Summary**
add [contenthash] support

**Does this PR introduce a breaking change?**
no

**Other information**

### CODE REVIEW

File: lib/JavascriptModulesPlugin.js [hunk 1]
```diff
@@ -124,6 +127,27 @@ class JavascriptModulesPlugin {
 						return result;
 					}
 				);
+				compilation.hooks.contentHash.tap("JavascriptModulesPlugin", chunk => {
+					const outputOptions = compilation.outputOptions;
+					const hashFunction = outputOptions.hashFunction;
```
Comment (michael-ciniawsky) [inline lib/JavascriptModulesPlugin.js]:
```js
const { 
  hashSalt, 
  hashDigest, 
  hashDigestLength, 
  hashFunction 
} = outputOptions;
```

File: lib/TemplatedPathPlugin.js [hunk 2]
```diff
@@ -7,6 +7,7 @@
 const REGEXP_HASH = /\[hash(?::(\d+))?\]/gi,
 	REGEXP_CHUNKHASH = /\[chunkhash(?::(\d+))?\]/gi,
 	REGEXP_MODULEHASH = /\[modulehash(?::(\d+))?\]/gi,
+	REGEXP_CONTENTHASH = /\[contenthash(?::(\d+))?\]/gi,
```
Comment (Kovensky) [inline lib/TemplatedPathPlugin.js]: Does `loader-utils` also need support for this; or does it just forward the call to webpack?

https://github.com/webpack/loader-utils#interpolatename

File: lib/JavascriptModulesPlugin.js [hunk 3]
```diff
@@ -72,6 +73,7 @@ class JavascriptModulesPlugin {
 							filenameTemplate,
 							pathOptions: {
 								noChunkHash: !useChunkHash,
+								contentHashType: "javascript",
```
Comment (Kovensky) [inline lib/JavascriptModulesPlugin.js]: I assume this means other plugins/loaders that don't produce javascript will need their own hashing system.

File: lib/TemplatedPathPlugin.js (hunk 2 above)
Comment (evilebottnawi) [reply lib/TemplatedPathPlugin.js]: @Kovensky :+1:

Comment (sokra) [reply lib/TemplatedPathPlugin.js]: This is unrelated to getName in loader-utils

File: lib/JavascriptModulesPlugin.js (hunk 3 above)
Comment (sokra) [reply lib/JavascriptModulesPlugin.js]: yep, I'll submit a PR to mini-css-extract-plugin where you can see an example how to use it.

**Pull request merged**